src/main/java/org/folio/am/service/ApplicationDescriptorsValidationService.java

```diff
@@ -0,0 +1,108 @@
+package org.folio.am.service;
+
+import static org.folio.am.utils.CollectionUtils.union;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang3.StringUtils;
+import org.folio.am.domain.dto.ApplicationDescriptor;
+import org.folio.am.domain.dto.ApplicationDto;
+import org.folio.am.domain.dto.Dependency;
+import org.folio.am.mapper.ApplicationDescriptorToDtoMapper;
+import org.folio.am.mapper.ApplicationEntityToDtoMapper;
+import org.semver4j.RangesListFactory;
+import org.springframework.stereotype.Service;
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class ApplicationDescriptorsValidationService {
+
+  private final ApplicationService applicationService;
+  private final ApplicationDescriptorToDtoMapper applicationDescriptorToDtoMapper;
+  private final ApplicationEntityToDtoMapper applicationEntityToDtoMapper;
+  private final DependenciesValidator dependenciesValidator;
+
+  public List<String> validateDescriptors(List<ApplicationDescriptor> descriptors) {
+    log.info("validateDescriptors:: validate descriptors ids {}", getDescriptorIdsAsStr(descriptors));
+    var applicationDtos = descriptors
+      .stream()
+      .map(applicationDescriptorToDtoMapper::convert)
+      .collect(Collectors.toCollection(LinkedHashSet::new));
+    var dependencyQueue = applicationDtos
+      .stream()
+      .map(ApplicationDto::getDependencies)
+      .flatMap(Collection::stream)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toCollection(LinkedList::new));
+    var visited = new HashSet<>();
+    while (!dependencyQueue.isEmpty()) {
+      var dependency = dependencyQueue.poll();
+      if (!visited.contains(dependency)) {
+        var applicationDtoOpt = getByLatestDependencyVersion(dependency, applicationDtos);
+        applicationDtoOpt.ifPresent(applicationDto -> {
+          applicationDtos.add(applicationDto);
+          dependencyQueue.addAll(applicationDto.getDependencies());
+          visited.add(dependency);
+        });
+      }
+    }
+    log.info("validateDescriptors:: validate applications including dependencies by ids {}",
+      getApplicationDtosIdsAsStr(applicationDtos));
+    dependenciesValidator.validate(new ArrayList<>(applicationDtos));
+    return applicationDtos
```

Comment:
Could you please explain why validate method returns some list of ids, shouldn't it throw an exception if something is wrong or just return `void` if everything is Ok

Reply:
Hi @dmtkachenko  I have updated the approach with more details. The response should provided info about successful applications.